File: src/scan_repair.py
import rospy
import numpy as np
from sensor_msgs.msg import LaserScan

scan_publisher = None

def convert_scan(data):
    global scan_publisher

    ignore_angle_min = -0.523599
    ignore_angle_max = 0.523599

    # 新しいLaserScanメッセージを作成します。
    filtered_scan = LaserScan()
    filtered_scan.header = data.header
    filtered_scan.angle_min = data.angle_min
    filtered_scan.angle_max = data.angle_max
    filtered_scan.angle_increment = data.angle_increment
    filtered_scan.time_increment = data.time_increment
    filtered_scan.scan_time = data.scan_time
    filtered_scan.range_min = data.range_min
    filtered_scan.range_max = data.range_max

    # 無視する範囲のデータを除外します。
    for i, angle in enumerate(np.arange(data.angle_min, data.angle_max, data.angle_increment)):
        if angle > ignore_angle_min and angle < ignore_angle_max:
            filtered_scan.ranges.append(float('nan'))
        else:
            filtered_scan.ranges.append(data.ranges[i])
            
    # フィルタリング後のデータをパブリッシュします。
    scan_publisher.publish(filtered_scan)


if __name__ == '__main__':
    rospy.init_node('scan_repair', log_level=rospy.DEBUG)

    rospy.Subscriber('/scan', LaserScan, convert_scan)
    scan_publisher = rospy.Publisher('/scan_front', LaserScan, queue_size=10)

    r = rospy.Rate(10)
    while not rospy.is_shutdown():
        r.sleep()

File: src/scan_repair.cpp
#include <ros/ros.h>
#include <sensor_msgs/LaserScan.h>
#include <vector>

ros::Publisher scan_publisher;

void convertScan(const sensor_msgs::LaserScan::ConstPtr& input_scan){
    const float max_angle = 1.570796;
    const float min_angle = -1.570796;

    // new scan data
    sensor_msgs::LaserScan filtered_scan;
    filtered_scan.header = input_scan->header;
    filtered_scan.angle_min = input_scan->angle_min;
    filtered_scan.angle_max = input_scan->angle_max;
    filtered_scan.angle_increment = input_scan->angle_increment;
    filtered_scan.time_increment = input_scan->time_increment;
    filtered_scan.scan_time = input_scan->scan_time;
    filtered_scan.range_min = input_scan->range_min;
    filtered_scan.range_max = input_scan->range_max;

    // filtere data
    int num_ranges = (input_scan->angle_max - input_scan->angle_min) / input_scan->angle_increment;
    filtered_scan.ranges.resize(num_ranges);
    filtered_scan.intensities.resize(num_ranges);

    //printf("increment:%f", input_scan->angle_increment);

    for (int i=0; i < num_ranges; i++){
        float angle = input_scan->angle_min + i * input_scan->angle_increment;
        if(angle > min_angle && angle < max_angle){
            filtered_scan.ranges[i] = std::numeric_limits<float>::infinity();
        } else {
            filtered_scan.ranges[i] = input_scan->ranges[i];
        }
        filtered_scan.intensities[i] = input_scan->intensities[i];
    }

    //publish
    scan_publisher.publish(filtered_scan);
}

int main(int argc, char** argv){
    ros::init(argc, argv, "scan_repair");
    ros::NodeHandle nh;

    scan_publisher = nh.advertise<sensor_msgs::LaserScan>("/scan", 10);
    ros::Subscriber scan_subscriber = nh.subscribe("/scan_raw", 10, convertScan);

    ros::spin();

    return 0;
}